code/src/main/java/com/onlinecontacttracing/servlets/authentication/CheckForCalendarAuthorizationServlet.java

```diff
@@ -0,0 +1,87 @@
+package com.onlinecontacttracing.authentication;
+
+import java.io.IOException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import com.google.gson.Gson;
+import java.util.ArrayList;
+
+import java.util.Collections;
+import com.onlinecontacttracing.authentication.CheckForCredentials;
+
+
+import com.google.api.client.auth.oauth2.Credential;
+import com.google.api.client.googleapis.auth.oauth2.GoogleAuthorizationCodeFlow;
+import com.google.api.client.googleapis.auth.oauth2.GoogleClientSecrets;
+import com.google.api.client.googleapis.javanet.GoogleNetHttpTransport;
+import com.google.api.client.http.javanet.NetHttpTransport;
+import com.google.api.client.json.JsonFactory;
+import com.google.api.client.json.jackson2.JacksonFactory;
+import com.google.api.services.gmail.Gmail;
+import com.google.api.services.gmail.GmailScopes;
+import com.google.api.services.gmail.model.Label;
+import com.google.api.services.gmail.model.ListLabelsResponse;
+import com.google.api.client.auth.oauth2.TokenResponse;
+import com.google.api.client.auth.oauth2.AuthorizationRequestUrl;
+import java.util.HashMap;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import java.util.Map;
+import com.fasterxml.jackson.core.type.TypeReference;
+import javax.servlet.RequestDispatcher;
+import com.google.api.services.calendar.CalendarScopes;
+import java.io.PrintWriter;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.security.GeneralSecurityException;
+import java.util.Collections;
+import java.util.List;
+import com.google.api.client.googleapis.auth.oauth2.GoogleIdToken.Payload;
+import java.util.Optional;
+
+@WebServlet("/check-for-calendar-credentials")
+public class CheckForCalendarAuthorizationServlet extends HttpServlet {
```

Comment:
Rather than having the same authorization logic duplicated in multiple places, let's prefer this code structure:
```
abstract class CheckForApiAuthorizationServlet extends HttpServlet {
  abstract List<String> getScopes();
  abstract void onSuccessfulLogin(Credential credential);
  @Override
  public void doPost(...) {
    // the code you have in doPost now, but using those abstract methods to abstract away the API specific information.
  }
}


@WebServlet("/check-for-calendar-credentials")
public class CheckForCalendarAuthorizationServlet extends CheckForApiAuthorizationServlet {
 // implement the Abstract methods above for calendar.
}

```